### ISSUE TITLE
Enhance Popup command for wheel gestures by directly selecting/switching items via scrolling

### ISSUE THREAD

**Issue by firemeteorxx** [Robbendebiene/Gesturefy#245]
Two things I love about the pop-up in the original FireGestures add-on:
1. Page history pop-up
2. Get Focus on the pop-up (at proper item given the context) and stop the gesture -- so that wheel operation could be used to operate the pop-up directly.

Can we do those as well in Gestruefy?

**Comment by Robbendebiene** [comment]
The first suggestion is described here: https://github.com/Robbendebiene/Gesturefy/issues/242; https://github.com/Robbendebiene/Gesturefy/issues/99

Can you please clarify your second suggestion a little bit more. Especially:

> stop the gesture

and
> (at proper item given the context)

**Comment by firemeteorxx** [comment]
Based on V1.2.0, the popup wheel gesture work in such way:
Once the action get triggered, a new context menu is shown -- not focused however.
If I keep scrolling the wheel, e.g. in different direction, the action will keep triggering, different menu will keep popping up.

My expectation, once the menu pops up, stops the gesture and focus on the menu immediately.
The focus is expected start from current tab / history. And further wheel scrolling can be used to navigate the menu.

Did I make it clear this time?

BTW, I'm surprised that on Linux Gesturefy stick at v1.1.7. Why the latest v1.2.0 is not shown up?

**Comment by Robbendebiene** [comment]
The popup is always focused, however this does not mean, that the scroll wheel will scroll the popups content if your mouse isn't over the popup. 

> If I keep scrolling the wheel, e.g. in different direction, the action will keep triggering, different menu will keep popping up.

This sounds like you are still holding the wheel trigger button (probably the left or right mouse button).
I guess you want to scroll through the popup by selecting the items one after another without moving your mouse over it and you also want to confirm an item from the list when you release your mouse button? Although I see some advantages of this technique, I'm personally against holding down the trigger button while you are scrolling through the popup. This would also require a different behavior for normal mouse gestures, since the trigger button is always released before the popup opens.
But maybe I still got you wrong? :)

> BTW, I'm surprised that on Linux Gesturefy stick at v1.1.7. Why the latest v1.2.0 is not shown up?

I'm currently on Linux with the latest Gesturefy version (1.2.0), you could try updating it via about:addons.

**Comment by firemeteorxx** [comment]
>This sounds like you are still holding the wheel trigger button (probably the left or right mouse button).
I guess you want to scroll through the popup by selecting the items one after another without moving your mouse over it and you also want to confirm an item from the list when you release your mouse button? 

I think you get my usage scenario correct. This is how I used it with the FireGesture add-on.
However, since it has been awhile, I couldn't remember exactly if the last step on 'confirm an item' is precise. It should be though, otherwise it won't be useful.

> This would also require a different behavior for normal mouse gestures, since the trigger button is always released before the popup opens.

Actually it does not work that way, whether it's intentional or by accident. You may have a try by yourself.

The rationale here is that 'wheel' by itself is NOT part of normal mouse gestures.
You can not chain 'wheel operation' in the middle of a longer gesture. (e.g. 'scroll down + scroll up + left + up == action A)
If you know a gesture is to be ended for definite, what's the point of following the 'usual behavior' here?
I really wish you do not fix the behavior towards what you described above...
I feel what I proposed is more useful. :-)

>I'm currently on Linux with the latest Gesturefy version (1.2.0), you could try updating it via about:addons

Please forget about it. It may has something to do with the old firefox v57.03 version.
After updating to v58 it prompts addon update as expected.

**Comment by Robbendebiene** [comment]
> Actually it does not work that way, whether it's intentional or by accident. You may have a try by yourself.

I guess you've misunderstood me. I was talking about performing a normal mouse gesture / the mouse gesture trigger button not the wheel gesture trigger button.

> The rationale here is that 'wheel' by itself is NOT part of normal mouse gestures.
> You can not chain 'wheel operation' in the middle of a longer gesture. (e.g. 'scroll down + scroll up + left + up == action A)
If you know a gesture is to be ended for definite, what's the point of following the 'usual behavior' here?

Currently I'm using the same actions for the normal mouse gestures, rocker and wheel gestures. So from an implementation point of view I would have to separate them, because the popup action won't work for normal mouse gestures in the way you described it above. Besides that I'm also a big fan of consistency. Do you know how FireGesture handled its popups for normal mouse gestures?

> I really wish you do not fix the behavior towards what you described above...

I may have missed it, but where did I mention any "fix" for this? (except from leaving it as it is) :D

**Comment by firemeteorxx** [comment]
Sorry for my misunderstanding~~ It's clear now.

>Do you know how FireGesture handled its popups for normal mouse gestures?

It's unfortunate but I only used pop-ups with wheel gesture -- the most handy scenario in my mind.

Here is my guess -- In terms of action activation, I think it's the wheel gesture that is different / inconsistent, not the pop-ups. 

For popups, I guess we could still define a selected-by-default item for consistency. It'll be less useful though. You may control the selection through arrow keys (e.g. up/down), but this defeats the motivation of mouse gesture -- minimize movement of hands.

From my test, Firefox native pop-up menus could react to arrow keys. 
But the one generated by Gesturefy does not.

**Comment by istrasci** [comment]
+1 for Page History popup.  One of my favourite FireGestures features.

**Comment by Reylak** [comment]
I agree with @firemeteorxx, I'm super happy to finally have a few pop up gestures but I also prefere the FireGesture's behavior. That is:

 * hold trigger button
 * scroll once in whichever direction
 * if it's a pop up, scrolling in any direction scrolls the popped up list
 * release the trigger button to validate the item

It's really smoother and quicker, it makes for a more agile experience.
